x-pack/platform/plugins/shared/lens/public/react_embeddable/initializers/initialize_edit.tsx

```diff
@@ -149,26 +150,39 @@ export function initializeEditApi(
     stateApi.updateSavedObjectId(newState.savedObjectId);
   };
 
-  // Wrap the getState() when inline editing and make sure that the filters in the attributes
-  // are properly injected with the correct references to avoid issues when saving/navigating to the full editor
-  const getStateWithInjectedFilters = () => {
+  /**
+   * Use the search context api here for filters for 2 reasons:
+   *  - the filters here have the correct references already injected
+   *  - the edit filters flow may change in the future and this is the right place to get the filters
+   */
+  const getFilters = ({ attributes }: LensRuntimeState): Filter[] =>
+    searchContextApi.filters$.getValue() ?? attributes.state.filters;
+
+  const convertVisualizationState = ({ attributes }: LensRuntimeState) => {
+    const visState = attributes.state.visualization;
+    const convertToRuntimeState =
+      startDependencies.visualizationMap[attributes.visualizationType ?? '']?.convertToRuntimeState;
+    if (!convertToRuntimeState) return visState;
+    return convertToRuntimeState(visState, attributes.state.datasourceStates);
+  };
```

Comment:
This is the partial solution for #207957 which is to apply the conversion at the highest level to avoid state and vis type incompatibilities.